File: source/clibCore/CPContentClass.vb
Imports Contensive.BaseClasses
Imports System.Runtime.InteropServices

Namespace Contensive.Core
    '
    '====================================================================================================
    '====================================================================================================
    ' comVisible to be activeScript compatible
    '
    <ComVisible(True)>
    <ComClass(CPContentClass.ClassId, CPContentClass.InterfaceId, CPContentClass.EventsId)>
    Public Class CPContentClass
        Inherits CPContentBaseClass
        Implements IDisposable
        '
#Region "COM GUIDs"
        Public Const ClassId As String = "D8D3D8F9-8459-46F7-B8AC-01B4DFAA4DB2"
        Public Const InterfaceId As String = "9B321DE5-D154-4EB1-B533-DBA2E5F2B5D2"
        Public Const EventsId As String = "6E068297-E09E-42C8-97B6-02DE591009DD"
#End Region
        '
        Private cp As CPClass
        Private cpCore As Contensive.Core.cpCoreClass
        Protected disposed As Boolean = False
        '
        '====================================================================================================
        '
        Friend Sub New(ByVal cpParent As CPClass)
            MyBase.New()
            cp = cpParent
            cpCore = cp.core
        End Sub
        '
        '====================================================================================================
        ' dispose
        '
        Protected Overridable Overloads Sub Dispose(ByVal disposing As Boolean)
            If Not Me.disposed Then
                Call appendDebugLog(".dispose, dereference main, csv")
                If disposing Then
                    '
                    ' call .dispose for managed objects
                    '
                    cpCore = Nothing
                    cp = Nothing
                End If
                '
                ' Add code here to release the unmanaged resource.
                '
            End If
            Me.disposed = True
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Function GetCopy(ByVal CopyName As String, Optional ByVal DefaultContent As String = "") As String
            GetCopy = cpCore.html_GetContentCopy(CopyName, DefaultContent, cpCore.user.userId, True, cpCore.user.user_isAuthenticated)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetCopy(ByVal CopyName As String, ByVal DefaultContent As String, ByVal personalizationPeopleId As Integer) As String
            GetCopy = cpCore.html_GetContentCopy(CopyName, DefaultContent, personalizationPeopleId, True, cpCore.user.user_isAuthenticated)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Sub SetCopy(ByVal CopyName As String, ByVal Content As String)
            Call cpCore.content_SetContentCopy(CopyName, Content)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Function GetAddLink(ByVal ContentName As String, ByVal PresetNameValueList As String, ByVal AllowPaste As Boolean, ByVal IsEditing As Boolean) As String
            Return cpCore.main_GetRecordAddLink2(ContentName, PresetNameValueList, AllowPaste, IsEditing)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetContentControlCriteria(ByVal ContentName As String) As String
            Return cpCore.db_GetContentControlCriteria(ContentName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetFieldProperty(ByVal ContentName As String, ByVal FieldName As String, ByVal PropertyName As String) As String
            Return cpCore.db_GetContentFieldProperty(ContentName, FieldName, PropertyName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetID(ByVal ContentName As String) As Integer
            Return cpCore.metaData.getContentId(ContentName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetProperty(ByVal ContentName As String, ByVal PropertyName As String) As String
            Return cpCore.db_GetContentProperty(ContentName, PropertyName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetDataSource(ByVal ContentName As String) As String
            Return cpCore.metaData.getContentDataSource(ContentName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetEditLink(ByVal ContentName As String, ByVal RecordID As String, ByVal AllowCut As Boolean, ByVal RecordName As String, ByVal IsEditing As Boolean) As String
            If True Then
                Return cpCore.main_GetRecordEditLink2(ContentName, RecordID, AllowCut, RecordName, IsEditing)
            Else
                Return ""
            End If
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetLinkAliasByPageID(ByVal PageID As Integer, ByVal QueryStringSuffix As String, ByVal DefaultLink As String) As String
            If True Then
                Return cpCore.main_GetLinkAliasByPageID(PageID, QueryStringSuffix, DefaultLink)
            Else
                Return ""
            End If
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetPageLink(ByVal PageID As Integer, Optional ByVal QueryStringSuffix As String = "", Optional ByVal AllowLinkAlias As Boolean = True) As String
            If True Then
                Return cpCore.main_GetPageLink3(PageID, QueryStringSuffix, AllowLinkAlias)
            Else
                Return ""
            End If
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetRecordID(ByVal ContentName As String, ByVal RecordName As String) As Integer
            Return cpCore.db_GetRecordID(ContentName, RecordName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetRecordName(ByVal ContentName As String, ByVal RecordID As Integer) As String
            Return cpCore.db_GetRecordName(ContentName, RecordID)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetTable(ByVal ContentName As String) As String
            Return cpCore.metaData.getContentTablename(ContentName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetTemplateLink(ByVal TemplateID As Integer) As String
            Return cpCore.main_GetTemplateLink(TemplateID)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function IsField(ByVal ContentName As String, ByVal FieldName As String) As Boolean
            Return cpCore.metaData_IsContentFieldSupported(ContentName, FieldName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function IsLocked(ByVal ContentName As String, ByVal RecordID As String) As Boolean 'Inherits CPContentBaseClass.IsLocked
            Return cpCore.workflow.isRecordLocked(ContentName, RecordID, 0)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function IsChildContent(ByVal ChildContentID As String, ByVal ParentContentID As String) As Boolean 'Inherits CPContentBaseClass.IsChildContent
            Return cpCore.metaData.isWithinContent(ChildContentID, ParentContentID)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function IsWorkflow(ByVal ContentName As String) As Boolean
            ' ****************************************************
            ' this needs to be added to cmc
            ' ****************************************************
            Return False
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Sub PublishEdit(ByVal ContentName As String, ByVal RecordID As Integer) 'Inherits CPContentBaseClass.PublishEdit
            Call cpCore.workflow.publishEdit(ContentName, RecordID, 0)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Sub SubmitEdit(ByVal ContentName As String, ByVal RecordID As Integer) 'Inherits CPContentBaseClass.SubmitEdit
            Call cpCore.workflow.submitEdit2(ContentName, RecordID, 0)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Sub AbortEdit(ByVal ContentName As String, ByVal RecordId As Integer) 'Inherits CPContentBaseClass.AbortEdit
            Call cpCore.workflow.abortEdit2(ContentName, RecordId, 0)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Sub ApproveEdit(ByVal ContentName As String, ByVal RecordId As Integer) 'Inherits CPContentBaseClass.ApproveEdit
            Call cpCore.workflow.approveEdit(ContentName, RecordId, 0)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Function getLayout(ByVal layoutName As String) As String
            Dim result As String = ""
            Try
                Dim cs As CPCSClass
                '
                cs = New CPCSClass(cp)
                cs.Open("layouts", "name=" & cp.Db.EncodeSQLText(layoutName), "id",, "layout")
                If cs.OK Then
                    result = cs.GetText("layout")
                End If
                Call cs.Close()
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in getLayout")
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function AddRecord(ByVal ContentName As String, ByVal recordName As String) As Integer
            Dim recordId As Integer = 0
            Try
                Dim cs As CPCSClass = cp.CSNew()
                If cs.Insert(ContentName) Then
                    cs.SetField("name", recordName)
                    recordId = cs.GetInteger("id")
                End If
                Call cs.Close()
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in AddRecord")
            End Try
            Return recordId
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function AddRecord(ContentName As String) As Integer
            Dim recordId As Integer = 0
            Try
                Dim cs As CPCSClass = cp.CSNew()
                If cs.Insert(ContentName) Then
                    recordId = cs.GetInteger("id")
                End If
                Call cs.Close()
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in AddRecord")
            End Try
            Return recordId
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Sub Delete(ByVal ContentName As String, ByVal SQLCriteria As String)
            Call cpCore.db.db_DeleteContentRecords(ContentName, SQLCriteria)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Sub DeleteRecords(ByVal ContentName As String, ByVal SQLCriteria As String)
            Call cpCore.db.db_DeleteContentRecords(ContentName, SQLCriteria)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Sub DeleteContent(ContentName As String)
            cpCore.metaData.deleteContent(ContentName)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Function AddContentField(ContentName As String, FieldName As String, FieldType As Integer) As Integer
            Dim builder As New coreBuilderClass(cpCore)
            Dim field As New coreMetaDataClass.CDefFieldClass()
            field.active = True
            field.adminOnly = False
            field.authorable = True
            field.blockAccess = False
            field.caption = FieldName
            field.contentId = cpCore.metaData.getContentId(ContentName)
            field.developerOnly = False
            field.editSortPriority = 9999
            field.editTabName = ""
            field.fieldTypeId = FieldType
            field.htmlContent = False
            field.indexColumn = 0
            field.indexSortDirection = 0
            field.indexSortOrder = 0
            field.indexWidth = 0
            field.installedByCollectionGuid = ""
            field.isBaseField = False
            field.lookupContentID = 0
            field.lookupContentName = ""
            field.lookupList = ""
            field.manyToManyContentID = 0
            field.ManyToManyContentName = ""
            field.manyToManyRuleContentID = 0
            field.ManyToManyRuleContentName = ""
            field.ManyToManyRulePrimaryField = ""
            field.ManyToManyRuleSecondaryField = ""
            field.MemberSelectGroupID = 0
            field.MemberSelectGroupName = ""
            field.nameLc = FieldName.ToLower()
            field.Password = False
            field.ReadOnly = False
            field.RedirectContentID = 0
            field.RedirectContentName = ""
            field.RedirectID = 0
            field.RedirectPath = ""
            field.Required = False
            field.Scramble = False
            field.TextBuffered = False
            field.UniqueName = False
            Return builder.metaData_VerifyCDefField_ReturnID(ContentName, field)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function AddContent(ContentName As String) As Integer
            Return AddContent(ContentName, ContentName.Replace(" "c, "").Replace(" "c, ""), "default")
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function AddContent(ContentName As String, sqlTableName As String) As Integer
            Return AddContent(ContentName, sqlTableName, "default")
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function AddContent(ContentName As String, sqlTableName As String, dataSource As String) As Integer
            Dim builder As New coreBuilderClass(cpCore)
            Return builder.metaData_CreateContent4(True _
                , dataSource _
                , sqlTableName _
                , ContentName _
                , False _
                , False _
                , True _
                , False _
                , "" _
                , "sort order" _
                , "name" _
                , False _
                , False _
                , False _
                , False _
                , False _
                , False _
                , "" _
                , 0 _
                , 0 _
                , 0 _
                , "" _
                , False _
                , "" _
                , True
                )
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetListLink(ContentName As String) As String
            Dim returnHtml As String = ""
            Try
                Dim adminUrl As String = cp.Site.GetText("adminUrl") & "?cid=" & cp.Content.GetID(ContentName)
                Dim encodedCaption As String = "List Records in " & cp.Utils.EncodeHTML(ContentName)

                returnHtml = returnHtml _
                    & "<a" _
                    & " class=""ccRecordEditLink"" " _
                    & " TabIndex=-1" _
                    & " href=""" & adminUrl & """" _
                    & "><img" _
                    & " src=""/ccLib/images/IconContentEdit.gif""" _
                    & " border=""0""" _
                    & " alt=""" & encodedCaption & """" _
                    & " title=""" & encodedCaption & """" _
                    & " align=""absmiddle""" _
                    & "></a>"
            Catch ex As Exception
                Call cp.core.handleExceptionAndRethrow(ex, "Unexpected error in GetListLink")
            End Try
            Return returnHtml
        End Function
        '
        '====================================================================================================
        ' append to logfile
        '
        Private Sub appendDebugLog(ByVal copy As String)
            'My.Computer.FileSystem.WriteAllText("c:\clibCpDebug.log", Now & " - cp.content, " & copy & vbCrLf, True)
        End Sub
        '
        '====================================================================================================
        ' testpoint
        '
        Private Sub tp(ByVal msg As String)
            'Call appendDebugLog(msg)
        End Sub
        '
        '====================================================================================================
        '
#Region " IDisposable Support "
        ' Do not change or add Overridable to these methods.
        ' Put cleanup code in Dispose(ByVal disposing As Boolean).
        Public Overloads Sub Dispose() Implements IDisposable.Dispose
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub
        Protected Overrides Sub Finalize()
            Dispose(False)
            MyBase.Finalize()
        End Sub
#End Region
    End Class
End Namespace

File: source/Processor/Views/CPContentClass.cs

using System;
using System.Collections.Generic;
using Contensive.Processor.Models.Db;
using Contensive.Processor.Controllers;
using Contensive.BaseClasses;
using Contensive.Addons.AdminSite.Controllers;
using Contensive.Processor.Models.Domain;
using Contensive.Processor.Exceptions;
using System.Data;

namespace Contensive.Processor {
    public class CPContentClass : CPContentBaseClass, IDisposable {
        //
        #region COM GUIDs
        public const string ClassId = "D8D3D8F9-8459-46F7-B8AC-01B4DFAA4DB2";
        public const string InterfaceId = "9B321DE5-D154-4EB1-B533-DBA2E5F2B5D2";
        public const string EventsId = "6E068297-E09E-42C8-97B6-02DE591009DD";
        #endregion
        //
        private CPClass cp { get; set; }
        //
        //====================================================================================================
        //
        public CPContentClass(CPClass cp) : base() {
            this.cp = cp;
        }
        //
        //====================================================================================================
        //
        public override int GetTableID(string tableName) {
            throw new NotImplementedException();
        }
        //
        //====================================================================================================
        //
        public override string GetCopy(string copyName, string DefaultContent) {
            return cp.core.html.getContentCopy(copyName, DefaultContent, cp.core.session.user.id, true, cp.core.session.isAuthenticated);
        }
        //
        public override string GetCopy(string copyName) {
            return cp.core.html.getContentCopy(copyName, "", cp.core.session.user.id, true, cp.core.session.isAuthenticated);
        }
        //
        //====================================================================================================
        //
        public override string GetCopy(string copyName, string defaultContent, int personalizationPeopleId) {
            return cp.core.html.getContentCopy(copyName, defaultContent, personalizationPeopleId, true, cp.core.session.isAuthenticated);
        }
        //
        //====================================================================================================
        //
        public override void SetCopy(string copyName, string content) {
            cp.core.html.setContentCopy(copyName, content);
        }
        //
        //====================================================================================================
        //
        public override string GetAddLink(string contentName, string presetNameValueList, bool allowPaste, bool isEditing) {
            string result = "";
            foreach (var link in AdminUIController.getRecordAddLink(cp.core, contentName, presetNameValueList, allowPaste, isEditing)) {
                result += link;
            }
            return result;
        }
        //
        //====================================================================================================
        //
        public override string GetContentControlCriteria(string contentName) {
            var meta = ContentMetadataModel.createByUniqueName(cp.core, contentName);
            if (meta != null) { return meta.legacyContentControlCriteria; }
            return string.Empty;
        }
        //
        //====================================================================================================
        //
        [Obsolete("deprecated, instead create contentMeta, lookup field and use property of field", false)]
        public override string GetFieldProperty(string contentName, string fieldName, string propertyName) {
            throw new GenericException("getContentFieldProperty deprecated, instead create contentMeta, lookup field and use property of field");
        }
        //
        //====================================================================================================
        //
        public override string GetDataSource(string contentName) {
            var meta = ContentMetadataModel.createByUniqueName(cp.core, contentName);
            if (meta != null) { return meta.dataSourceName; }
            return string.Empty;
        }
        //
        //====================================================================================================
        //
        public override string GetEditLink(string contentName, string recordID, bool allowCut, string recordName, bool isEditing) {
            return AdminUIController.getRecordEditLink(cp.core, contentName, GenericController.encodeInteger(recordID), allowCut, recordName, isEditing);
        }
        //
        //====================================================================================================
        //
        public override string GetLinkAliasByPageID(int pageID, string queryStringSuffix, string defaultLink) {
            return LinkAliasController.getLinkAlias(cp.core, pageID, queryStringSuffix, defaultLink);
        }
        //
        //====================================================================================================
        //
        public override string GetPageLink(int pageID, string queryStringSuffix, bool allowLinkAlias) {
            return PageContentController.getPageLink(cp.core, pageID, queryStringSuffix, allowLinkAlias, false);
        }
        //
        public override string GetPageLink(int pageID, string queryStringSuffix) {
            return PageContentController.getPageLink(cp.core, pageID, queryStringSuffix, true, false);
        }
        //
        public override string GetPageLink(int pageID) {
            return PageContentController.getPageLink(cp.core, pageID, "", true, false);
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a record id from its unique name. If a duplicate exists, the first ordered by id is returned
        /// </summary>
        /// <param name="contentName"></param>
        /// <param name="recordName"></param>
        /// <returns></returns>
        public override int GetRecordID(string contentName, string recordName) {
            return MetadataController.getRecordIdByUniqueName(cp.core, contentName, recordName);
        }
        //
        //====================================================================================================
        /// <summary>
        /// returns the matching record name if a match is found, otherwise blank. Does NOT validate the record.
        /// </summary>
        /// <param name="contentName"></param>
        /// <param name="recordID"></param>
        /// <returns></returns>
        public override string GetRecordName(string contentName, int recordID) {
            var meta = ContentMetadataModel.createByUniqueName(cp.core, contentName);
            if (meta == null) { return string.Empty; }
            return meta.getRecordName(cp.core, recordID);
        }
        //
        //====================================================================================================
        //
        public override string GetTable(string contentName) {
            var meta = ContentMetadataModel.createByUniqueName(cp.core, contentName);
            if (meta == null) { return string.Empty; }
            return meta.tableName;
        }
        //
        //====================================================================================================
        //
        public override bool IsField(string contentName, string fieldName) {
            var contentMetadata = Models.Domain.ContentMetadataModel.createByUniqueName(cp.core, contentName);
            return (contentMetadata == null) ? false : contentMetadata.containsField(cp.core, fieldName);
        }
        //
        //====================================================================================================
        //
        public override bool IsLocked(string contentName, string recordId) {
            var contentTable = TableModel.createByContentName(cp.core, contentName);
            if (contentTable != null) return WorkflowController.isRecordLocked(cp.core, contentTable.id, GenericController.encodeInteger(recordId));
            return false;
        }
        //
        //====================================================================================================
        //
        public override bool IsChildContent(string childContentID, string parentContentID) {
            var parentMetadata = ContentMetadataModel.create(cp.core, parentContentID);
            return (parentMetadata == null) ? false : parentMetadata.isParentOf(cp.core, GenericController.encodeInteger(childContentID));
        }
        //
        //====================================================================================================
        //
        public override string getLayout(string layoutName) {
            string result = "";
            try {
                using (var cs = new CsModel(cp.core)) {
                    cs.open("layouts", "name=" + DbController.encodeSQLText(layoutName), "id", false, cp.core.session.user.id, "layout");
                    if (cs.ok()) {
                        result = cs.getText("layout");
                    }
                }
            } catch (Exception ex) {
                LogController.handleError(cp.core, ex);
                throw;
            }
            return result;
        }
        //
        //====================================================================================================
        //
        public override int AddRecord(string contentName, string recordName) {
            int recordId = 0;
            try {
                CsModel cs = new CsModel(cp.core);
                if (cs.insert(contentName)) {
                    cs.set("name", recordName);
                    recordId = cs.getInteger("id");
                }
                cs.close();
            } catch (Exception ex) {
                LogController.handleError(cp.core, ex);
                throw;
            }
            return recordId;
        }
        //
        //====================================================================================================
        //
        public override int AddRecord(string contentName) {
            int result = 0;
            try {
                CsModel cs = new CsModel(cp.core);
                if (cs.insert(contentName)) {
                    result = cs.getInteger("id");
                }
                cs.close();
            } catch (Exception ex) {
                LogController.handleError(cp.core, ex);
                throw;
            }
            return result;
        }
        //
        //====================================================================================================
        //
        public override void Delete(string contentName, string sqlCriteria) {
            MetadataController.deleteContentRecords(cp.core, contentName, sqlCriteria);
        }
        //
        //====================================================================================================
        //
        public override void DeleteContent(string contentName) {
            ContentModel.delete(cp.core, Models.Domain.ContentMetadataModel.getContentId(cp.core, contentName));
        }
        //
        //====================================================================================================
        //
        public override int AddContentField(string contentName, string fieldName, CPContentBaseClass.fileTypeIdEnum fieldType) {
            var contentMetadata = ContentMetadataModel.createByUniqueName(cp.core, contentName);
            var fieldMeta = ContentFieldMetadataModel.createDefault(cp.core, fieldName, fieldType);
            contentMetadata.verifyContentField(cp.core, fieldMeta, false);
            return fieldMeta.id;
        }
        //
        public override int AddContentField(string contentName, string fieldName, int fieldTypeId)
            => AddContentField(contentName, fieldName, (CPContentBaseClass.fileTypeIdEnum)fieldTypeId);
        //
        //====================================================================================================
        //
        public override int AddContent(string contentName) {
            return AddContent(contentName, contentName.Replace(' '.ToString(), "").Replace(' '.ToString(), ""), "default");
        }
        //
        //====================================================================================================
        //
        public override int AddContent(string contentName, string sqlTableName) {
            return AddContent(contentName, sqlTableName, "default");
        }
        //
        //====================================================================================================
        //
        public override int AddContent(string contentName, string sqlTableName, string dataSourceName) {
            var tmpList = new List<string> { };
            DataSourceModel dataSource = DataSourceModel.createByUniqueName(cp.core, dataSourceName, ref tmpList);
            return ContentMetadataModel.verifyContent_returnId(cp.core, new Models.Domain.ContentMetadataModel() {
                dataSourceName = dataSource.name,
                tableName = sqlTableName,
                name = contentName
            });
        }
        //
        //====================================================================================================
        //
        public override string GetListLink(string contentName) {
            return AdminUIController.getIconEditAdminLink(cp.core, Models.Domain.ContentMetadataModel.createByUniqueName(cp.core, contentName));
        }
        //
        //====================================================================================================
        //
        public override int GetID(string ContentName) {
            var content = ContentModel.createByUniqueName(cp.core, ContentName);
            if (content != null) return content.id;
            return 0;

        }
        //
        //====================================================================================================
        // deprecated
        //
        [Obsolete("Use AddRecord( string contentName ) ", false)]
        public override int AddRecord(object ContentName) {
            return AddRecord(cp.Utils.EncodeText(ContentName));
        }
        //
        [Obsolete("workflow editing is deprecated", false)]
        public override bool IsWorkflow(string ContentName) {
            //
            // -- workflow no longer supported (but may come back)
            return false;
        }
        //
        [Obsolete("workflow editing is deprecated", false)]
        public override void PublishEdit(string ContentName, int RecordID) {
            // Call WorkflowController.publishEdit(ContentName, RecordID, 0)
        }
        //
        [Obsolete("workflow editing is deprecated", false)]
        public override void SubmitEdit(string ContentName, int RecordID) {
            //Call WorkflowController.submitEdit2(ContentName, RecordID, 0)
        }
        //
        [Obsolete("workflow editing is deprecated", false)]
        public override void AbortEdit(string ContentName, int RecordId) {
            // Call WorkflowController.abortEdit2(ContentName, RecordId, 0)
        }
        //
        [Obsolete("workflow editing is deprecated", false)]
        public override void ApproveEdit(string ContentName, int RecordId) {
            //Call WorkflowController.approveEdit(ContentName, RecordId, 0)
        }
        //
        [Obsolete("Deprecated, template link is not supported", false)]
        public override string GetTemplateLink(int TemplateID) {
            return "";
        }
        //
        [Obsolete("Deprecated, access model properties instead", false)]
        public override string GetProperty(string ContentName, string PropertyName) {
            var contentMetadata = Models.Domain.ContentMetadataModel.createByUniqueName(cp.core, ContentName);
            return contentMetadata.getContentProperty(cp.core, PropertyName);
        }
        //
        //
        #region  IDisposable Support 
        //
        //====================================================================================================
        //
        protected virtual void Dispose(bool disposing) {
            if (!this.disposed) {
                if (disposing) {
                    //
                    // call .dispose for managed objects
                    //
                }
                //
                // Add code here to release the unmanaged resource.
                //
            }
            this.disposed = true;
        }
        protected bool disposed { get; set; } = false;
        // Do not change or add Overridable to these methods.
        // Put cleanup code in Dispose(ByVal disposing As Boolean).
        public void Dispose() {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        ~CPContentClass() {
            Dispose(false);
        }
        #endregion
    }
}
